src/frontend/src/expr/utils.rs

```diff
@@ -229,6 +241,51 @@ impl ExprRewriter for NotPushDown {
     }
 }
 
+pub fn extract_common_factor(expr: ExprImpl) -> Vec<ExprImpl> {
+    // For example, we assume that expr is (A & B & C & D) | (B & C & D) | (C & D & E)
+    let disjunctions: Vec<ExprImpl> = to_disjunctions(expr);
+
+    // if `expr` can not be divided into several disjunctions...
+    if disjunctions.len() == 1 {
+        return disjunctions;
+    }
+
+    // extract (A & B & C & D) | (B & C & D) | (C & D & E)
+    // into [[A, B, C, D], [B, C, D], [C, D, E]]
+    let mut disjunctions: Vec<Vec<_>> = disjunctions
+        .into_iter()
+        .map(|x| to_conjunctions(x).into_iter().collect())
+        .collect();
+    let mut last = disjunctions.pop().unwrap();
+    // now greatest_common_factor == [C, D]
+    let greatest_common_divider: Vec<_> = last
+        .drain_filter(|factor| disjunctions.iter().all(|expr| expr.contains(factor)))
+        .collect();
+    disjunctions.push(last);
```

Comment:
```suggestion
    let mut last = disjunctions.last_mut().unwrap();
    // now greatest_common_factor == [C, D]
    let greatest_common_divider: Vec<_> = last
        .drain_filter(|factor| disjunctions.iter().all(|expr| expr.contains(factor)))
        .collect();
```

Reply:
> `pop+push=last_mut`

There are some small differences. `let last = some_vec.pop().unwrap()` will transfers the ownership to `last`, while `last_mut` only takes a mutable reference. We cannot take ownership of a `Vec` element unless we remove it from the `Vec`. That's why `pop+push` is used. I agree with you that it is a bit strange. It would be nice if there were other better ways of doing this.

Reply:
You are right. I missed the usage of `disjunctions` inside the closure

Reply:
```rust
    let (last, remaining) = disjunctions.split_last_mut().unwrap();
    // now greatest_common_factor == [C, D]
    let greatest_common_divider: Vec<_> = last
        .drain_filter(|factor| remaining.iter().all(|expr| expr.contains(factor)))
        .collect();
```
🤪

Reply:
And in the following line we may also use `remaining` instead of `disjunctions`?

Reply:
That's great idea!  Done in 45571d43b01b139446dcf638aba3dc3ec40e6e73